loadtest/control/simulcontroller/actions.go

```diff
@@ -1017,3 +1056,92 @@ func searchGroupChannels(u user.User) control.UserActionResponse {
 		return control.UserActionResponse{Info: fmt.Sprintf("found %d channels", len(channels))}
 	})
 }
+
+func createPrivateChannel(u user.User) control.UserActionResponse {
+	team, err := u.Store().RandomTeam(store.SelectMemberOf)
+	if err != nil {
+		return control.UserActionResponse{Err: control.NewUserError(err)}
+	}
+
+	channelId, err := u.CreateChannel(&model.Channel{
+		Name:   model.NewId(),
+		TeamId: team.Id,
+		Type:   "P",
+	})
+	if err != nil {
+		return control.UserActionResponse{Err: control.NewUserError(err)}
+	}
+
+	// This is a series of calls made by the webapp client
+	// when opening the `Add Members` dialog.
+	if err := u.GetUsersInChannel(channelId, 0, 100); err != nil {
+		return control.UserActionResponse{Err: control.NewUserError(err)}
+	}
+
+	if err := u.GetChannelMembers(channelId, 0, 50); err != nil {
+		return control.UserActionResponse{Err: control.NewUserError(err)}
+	}
+
+	ids, err := u.GetUsersNotInChannel(team.Id, channelId, 0, 100)
+	if err != nil {
+		return control.UserActionResponse{Err: control.NewUserError(err)}
+	}
+
+	// we pick up to 4 users to add to the channel.
+	for _, id := range pickIds(ids, 1+rand.Intn(4)) {
+		if err := u.AddChannelMember(channelId, id); err != nil {
```

Comment:
We are not doing this for creating public channels. Is that because users will join the channel by themselves?

Reply:
Yes. Ideally we should have an "invite action" to invite users to channel we created and also a "leave channel" one so that they don't grow indefinitely.

> I might have forgotten some details about the algo, but how is changing from int to float helping?

The algorithm is more or less unchanged. The advantage of using float is that if I add an action I can always find a small enough number to fit that in. Before, if the action rate was less than 1 I had to multiply everything else by some constant to make sure the frequencies remained proportional to each other.